src/ledger/LedgerStateCache.h

```diff
@@ -0,0 +1,82 @@
+#pragma once
+
+// Copyright 2025 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "ledger/LedgerTxn.h"
+#include "xdr/Stellar-ledger-entries.h"
+#include <optional>
+#include <shared_mutex>
+#include <unordered_set>
+
+namespace stellar
+{
+// TODO Add ifdef
+struct LedgerEntryHash
+{
+    size_t operator()(LedgerEntry const& entry) const;
+};
+
+struct LedgerEntryEqual
+{
+    bool
+    operator()(LedgerEntry const& a, LedgerEntry const& b) const
+    {
+        return LedgerEntryKey(a) == LedgerEntryKey(b);
+    }
+};
+
+class LedgerManagerImpl;
+class PopulateLedgerCacheWork;
+
+class LedgerStateCache
+{
+    enum class Mode
+    {
+        ALL_ENTRIES,
+        SOROBAN_ONLY
+    };
+
+  private:
+    Mode mMode;
+    std::unordered_set<LedgerEntry, LedgerEntryHash, LedgerEntryEqual> mState;
```

Comment:
C++ 20 supports heterogenous unordered_set, but alas, C++17 does not. We might wanna benchmark this implementation where we create throw away dummy keys against just a plain old std::set with heterogenous lookups, set may be better.